dhall/src/Dhall/Import/Types.hs

```diff
@@ -5,56 +5,105 @@
 import Dhall.Binary (StandardVersion(..))
 import Dhall.Context (Context)
 import Dhall.Core
-  ( Directory (..)
-  , Expr
+  ( Chunks(..)
+  , Directory (..)
+  , Expr(..)
   , File (..)
   , FilePrefix (..)
-  , Import (..)
-  , ImportHashed (..)
   , ImportMode (..)
-  , ImportType (..)
   , ReifiedNormalizer(..)
+  , URL(..)
   )
 import Dhall.Parser (Src)
 import Dhall.TypeCheck (X)
 import Lens.Family (LensLike')
-import System.FilePath (isRelative, splitDirectories)
+import System.FilePath (isRelative, splitDirectories, takeFileName, takeDirectory)
 
+import qualified Crypto.Hash
 import qualified Dhall.Binary
 import qualified Dhall.Context
-import qualified Data.Map      as Map
+import qualified Dhall.Map
+import qualified Data.Map
 import qualified Data.Text
 
-data Resolved = Resolved
-    { resolvedExpression :: Expr Src Import
-    -- ^ Expression with its immediate imports resolved
-    , newImport          :: Import
-    -- ^ New import to use in place of the original import for chaining
-    --   downstream imports
-    }
+-- HTTP headers
+type Headers = [(Text, Text)]
 
--- | `parent` imports (i.e. depends on) `child`
-data Depends = Depends { parent :: Import, child :: Import }
+-- non-cyclic, binary headers, canonical
+data ResolvedImportType
+    = ResolvedLocal FilePrefix File
+      -- ^ Like a normal file import, but not resolved relative to the import stack
+    | ResolvedRemote URL (Maybe Headers)
+      -- ^ URL (sans headers) plus normalised headers
+    | ResolvedEnv Text  -- ^ same as plain Env
+    | ResolvedMissing  -- ^ same as plain Missing
+    deriving (Eq, Ord)
 
--- | State threaded throughout the import process
-data Status m = Status
-    { _stack :: NonEmpty Import
-    -- ^ Stack of `Import`s that we've imported along the way to get to the
-    -- current point
+instance Pretty ResolvedImportType where
+    pretty (ResolvedLocal prefix file) =
+        pretty prefix <> pretty file
+
+    pretty (ResolvedRemote url Nothing) = pretty url
+
+    pretty (ResolvedRemote url (Just headers)) =
+        pretty url <> " using " <> pretty headersExpr
+      where
+        headersExpr = ListLit Nothing (foldMap (pure . headerExpr) headers)
+        headerExpr :: (Text, Text) -> Expr Src X
+        headerExpr (key, value) =
+            let keyExpr = TextLit (Chunks [] key)
+                valueExpr = TextLit (Chunks [] value)
+            in RecordLit (Dhall.Map.fromList [("mapKey", keyExpr), ("mapValue", valueExpr)])
+
+    pretty (ResolvedEnv env) = "env:" <> pretty env
 
-    , _graph :: [Depends]
-    -- ^ Graph of all the imports visited so far, represented by a list of
-    --   import dependencies.
+    pretty ResolvedMissing = "missing"
 
-    , _cache :: Map Import (Expr Src X)
+
+type SemanticHash = Crypto.Hash.Digest Crypto.Hash.SHA256
+type SemisemanticHash = Crypto.Hash.Digest Crypto.Hash.SHA256
+
+data ResolvedImport = ResolvedImport (Maybe SemanticHash) ResolvedImportType ImportMode
```

Comment:
This boils down to my decision to represent 'resolved' headers as `(Text,Text)` pairs rather than `Expr Src Import`. I felt the latter type was a bit too big.

This was mainly to help me structure and understand the code while figuring things out. The decision is completely orthogonal to 'semi-semantic' hashing!

Reply:
I think we should share as much logic between the two types as possible, since it will simplify the remaining code a lot if we do.  So perhaps either use the same header type for both or parametrize the `ImportType` type on the type of the headers